beater/server_test.go

```diff
@@ -228,6 +228,77 @@ func TestServerNoContentType(t *testing.T) {
 	assert.Equal(t, http.StatusBadRequest, res.StatusCode, body(t, res))
 }
 
+func TestServerSourcemapElasticsearch(t *testing.T) {
+	cases := []struct {
+		expected     []string
+		config       m
+		outputConfig m
+	}{
+		{
+			expected: nil,
+			config:   m{},
+		},
+		{
+			// source_mapping.elasticsearch.hosts set
+			expected: []string{"localhost:5200"},
+			config: m{
+				"frontend": m{
+					"enabled": "true",
+					"source_mapping.elasticsearch.hosts": []string{"localhost:5200"},
+				},
+			},
+		},
+		{
+			// source_mapping.elasticsearch.hosts not set, elasticsearch.enabled = true
+			expected: []string{"localhost:5201"},
+			config: m{
+				"frontend": m{
+					"enabled": "true",
+				},
+			},
+			outputConfig: m{
+				"elasticsearch": m{
+					"enabled": true,
+					"hosts":   []string{"localhost:5201"},
+				},
+			},
+		},
+		{
+			// source_mapping.elasticsearch.hosts not set, elasticsearch.enabled = false
+			expected: nil,
+			config: m{
+				"frontend": m{
+					"enabled": "true",
+				},
+			},
+			outputConfig: m{
+				"elasticsearch": m{
+					"enabled": false,
+					"hosts":   []string{"localhost:5202"},
+				},
+			},
+		},
+	}
+	for _, testCase := range cases {
+		ucfg, err := common.NewConfigFrom(testCase.config)
+		if !assert.NoError(t, err) {
+			continue
+		}
+
+		var beatConfig beat.BeatConfig
+		ocfg, err := common.NewConfigFrom(testCase.outputConfig)
+		if !assert.NoError(t, err) {
+			continue
+		}
+		beatConfig.Output.Unpack(ocfg)
+		apm, teardown, err := setupServer(t, ucfg, &beatConfig)
+		if assert.NoError(t, err) {
+			assert.Equal(t, testCase.expected, apm.smapElasticsearchHosts())
```

Comment:
I think I am not seeing the point of this test. You asserting that `smapElasticsearchHosts` returns the expected thing, buy `smapElasticsearchHosts` itself is a test helper function only used here?

Reply:
The fallback from `source_mapping.elasticsearch` to `output.elasticsearch` only happens in `beater.New` - `smapElasticsearchHosts` is a test-only function that allows access to the final configuration of `Frontend.SourceMapping`.  These tests create a new beater to test that the fallback is happening (and not happening) correctly.

Reply:
ok, thanks. a comment would help